numpy/core/__init__.py

```diff
@@ -134,12 +135,31 @@ def _ufunc_reduce(func):
     return func.__name__
 
 
+def _DType_reconstruct(scalar_type):
+    # This is a work-around to pickle type(np.dtype(np.float64)), etc.
+    # and it should eventually be replaced with a better solution, e.g. when
+    # DTypes become HeapTypes.
+    return type(dtype(scalar_type))
+
+
+def _DType_reduce(DType):
+    # To pickle a DType without having to add top-level names, pickle the
+    # scalar type for now (and assume that reconstruction will be possible).
+    if DType is dtype:
+        return "dtype"  # must pickle `np.dtype` as a singleton.
```

Comment:
There was a subtlety/issue here. the change would have made `np.dtype` pickle in a way that older NumPy versions cannot unpickle. But when we pickle `np.dtype("f")` for example, we also pickle `np.dtype` to unpickle it.

That would have meant that any pickle containing a dtype instance (not class) would not have been unpickable on older NumPy versions.  For classes we cannot avoid that, but I am pretty sure that it would be inconvenient for the instances.